text/0000-refs.md

```diff
@@ -0,0 +1,209 @@
+- [Svelte: `bind:this`](https://svelte.dev/docs#template-syntax-element-directives-bind-this)
+- [Lit: `createRef()`](https://lit.dev/docs/api/directives/#ref)
+- [Angular: `elementRef`](https://angular.io/api/core/ElementRef)
+- [Stencil: `ref`](https://stenciljs.com/docs/templating-jsx#getting-a-reference-to-a-dom-element)
+
+By building on these conventions, we offer a simple and ergonomic way to solve the problem of locating elements at runtime.
+
+## Detailed design
+
+The basic design of refs is the use of the `lwc:ref` directive in templates:
+
+```html
+<template>
+    <div lwc:ref="foo"></div>
+</template>
+```
+
+...which can be accessed at runtime in a component using `this.refs`:
+
+```js
+export default class extends LightningElement {
+  renderedCallback() {
+    console.log(this.refs.foo); // `<div>` DOM node
+  }
+}
+```
+
+### `lwc:ref`
+
+`lwc:ref` can only refer to a static string. For instance, this will throw a template compiler error:
+
+```html
+<template>
+    <div lwc:ref={dynamic}></div> <!-- Not allowed! -->
+</template>
+```
+
+`lwc:ref` cannot be applied to `<template>`s, or to `<slot>`s in light DOM:
+
+```html
+<template>
+    <template if:true={foo} lwc:ref="foo"></template> <!-- Not allowed! -->
+</template>
+```
+
+```html
+<template lwc:render-mode="light">
+    <slot lwc:ref="foo"></slot> <!-- Not allowed! -->
+</template>
+```
+
+### `this.refs`
+
+`this.refs` is a plain object, with string keys and DOM node values. If a ref does not exist, then `this.refs["unknown"]` returns `undefined`.
+
+The object is read-only, and attempting to add, modify, or delete its properties from within the component will result in a runtime error.
+
+In the case of shadow DOM components, `this.refs` refers to elements inside of the shadow DOM. Whereas with light DOM components, it refers to elements inside of the light DOM. This makes sense, since in either case, the element must be defined in the template owned by that component in order to be accessible with `this.refs`.
+
+### Multiple templates
+
+In the case of multiple templates, the same rules as for `this.template.querySelector()` or `this.querySelector()` apply. Once the new template is rendered, `this.refs` refers to elements defined in the new template:
+
+```js
+import a from './a.html';
+import b from './b.html';
+
+export default class extends LightningElement {
+  count = 0;
+  
+  render() {
+    return this.count % 2 === 0 ? a : b;
+  }
+  
+  renderedCallback() {
+    console.log(this.refs);
+  }
+  
+  increment () {
+    this.count++;
+  }
+}
+```
+
+```js
+// Logs `this.refs` for a.html
+const cmp = createElement('x-component', { is: Component });
+
+// Logs `this.refs` for b.html
+cmp.increment();
+
+// Logs `this.refs` for a.html
+cmp.increment();
+```
+
+### Duplicate `lwc:ref`s
+
+If duplicate `lwc:ref`s are defined in the same template:
+
+```html
+<template>
+    <div lwc:ref="foo"></div>
+    <div lwc:ref="foo"></div>

```

Comment:
This example is obviously wrong. The template compiler can warn whenever the same `ref` is used twice is the same scope (same if and for block).

Reply:
My point, though, is: why should the template compiler bother? If a template has duplicates, it's not a big problem, as long as we have consistent ordering for which one "wins." Trying to hunt down every "incorrect" duplicate may be more trouble than it's worth.

I also noted that, in this same scenario, none of Svelte/Vue/React threw an error; they just had the last one win.